hop/robust_publisher.py

```diff
@@ -0,0 +1,633 @@
+                key = _ensure_bytes_like(header[0])
+                value = _ensure_bytes_like(header[1])
+                rbody.write(PublicationJournal.encode_int(len(key)))  # header key length
+                rbody.write(key)  # header key
+                rbody.write(PublicationJournal.encode_int(len(value)))  # header value length
+                rbody.write(value)  # header value
+        else:
+            rbody.write(PublicationJournal.encode_int(0))  # no headers
+        self._write_record(PublicationJournal.msg_record_type, rbody.getvalue())
+
+        self.messages_to_send[sequence_number] = (message, headers)
+        self.message_counter += 1
+        return sequence_number
+
+    def has_messages_to_send(self):
+        """Check whether there are messages queued for sending (which have either not been sent at
+            all, or for which all sending attempts so far have failed, causing them to be requeued).
+        """
+        return len(self.messages_to_send) > 0
+
+    def get_next_message_to_send(self):
+        """Fetch the next mesage which should be sent
+
+        Returns:
+            The next message in the form of a tuple of (seqeunce number, message, message headers),
+            or None if there are no messages currently needing to be sent.
+        """
+        if len(self.messages_to_send) == 0:
+            return None
+        result = self.messages_to_send.pop_highest_priority()
+        self.maybe_sent_messages[result[0]] = result[1]
+        # rearrange to be more friendly to downstream code
+        return (result[0], result[1][0], result[1][1])
+
+    def has_messages_in_flight(self):
+        """Check whether there are messages for which a sending attempt has been started,
+            but has not yet conclusively succeeded or failed
+        """
+        return len(self.maybe_sent_messages) > 0
+
+    def mark_message_sent(self, sequence_number):
+        """Mark a message as successfully sent, and removes it from further consideration.
+
+        Truncates and restarts the backing journal file if the number of messages in-flight and
+        waiting to be sent falls to zero, and restarts the sequence number assignment sequence.
+
+        Raises:
+            RuntimeError: If no message with the specifed sequence number is currently recorded as
+                being in-flight.
+        """
+        if sequence_number not in self.maybe_sent_messages:
+            raise RuntimeError("No record of message with sequence number " + str(sequence_number)
+                               + " being queued for sending")
+        self._write_record(PublicationJournal.sent_record_type,
+                           PublicationJournal.encode_int(sequence_number))
+        del self.maybe_sent_messages[sequence_number]
+        if len(self.messages_to_send) == 0 and len(self.maybe_sent_messages) == 0:
+            # take this opportunity to garbage collect the journal file
+            self.journal.close()
```

Comment:
The way I had thought about it was that the journal should be kept around unless we can prove we don't need I anymore. In general, one would like to garbage collect all pairs of matching (message-to-send, message-sent) records, but keeping the format simple and append-only meant that this can only be done when all queued messages are known to have been sent. It might be possible to do some somehwat fancier things (let the journal split into multiple files, implement it as a riing buffer, etc.) which would allow finer grained clean up, or clean up at other times, but off the top of my head, when we know we've sent a message is the only state change in the current scheme when we know we can clean up.